feat: add import capabilities to cdktf (#2972)
hashicorp/terraform-cdk: website/docs/cdktf/concepts/resources.mdx

If you need to ensure a condition is met either before or after a resource was created you can specify [conditions](/terraform/language/expressions/custom-conditions#preconditions-and-postconditions).
To add one configure the `lifecycle` key on your resource with an object containing a `precondition` and / or a `postcondition`. These keys take a list of conditions with a `condition` key containing a Terraform Expression to be evaluated and an `errorMessage` key containing a string to be displayed if the condition is not met.

## Importing Resources

If you have existing resources that you want to manage with CDKTF, you can import them into your CDKTF application. The best way to do this is using the [`import` block feature](/terraform/language/import) of Terraform >= 1.5. You can do this in CDKTF either with a specified configuration or without.

### How To Import

> To import a resource, first instantiate an instance of the resource type you wish to import– in our case we'll be using an S3Bucket. No configuration is explicitly needed. You then call the `importFrom` method on the resource object. This method takes the ID of the resource to be imported as the first argument and the provider as an optional second. The provider is only required if you have multiple providers of the same type in your configuration.

Reviewer: This got probably lost 🙈
```suggestion
To import a resource, first instantiate an instance of the resource type you wish to import – in our case we'll be using an S3Bucket. No configuration is explicitly needed. You then call the `importFrom` method on the resource object. This method takes the ID of the resource to be imported as the first argument and the provider as an optional second. The provider is only required if you have multiple providers of the same type in your configuration.
```

Reviewer 2: Whoops, included 👍

If you need to ensure a condition is met either before or after a resource was created you can specify [conditions](/terraform/language/expressions/custom-conditions#preconditions-and-postconditions).
To add one configure the `lifecycle` key on your resource with an object containing a `precondition` and / or a `postcondition`. These keys take a list of conditions with a `condition` key containing a Terraform Expression to be evaluated and an `errorMessage` key containing a string to be displayed if the condition is not met.

## Importing Resources

If you have existing resources that you want to manage with CDKTF, you can import them into your CDKTF application. The best way to do this is using the [`import` block feature](/terraform/language/import) of Terraform >= 1.5. You can do this in CDKTF either with a specified configuration or without, which will trigger the configuration generation.

### Importing with Configuration

> To import a resource with configuration, you write your resource configuration as you would normally. Then you call the `importFrom` method on the resulting resource object. This method takes the resource ID as the first argument and the provider as an optional second. The provider is only required if you have multiple providers of the same type in your configuration.

Reviewer: Upon reading this: Why don't we just take the provider that's passed to the resource and get rid of the second optional argument for `importFrom`? Would make it simpler for our users I think.

                           apply_server_side_encryption_by_default {
                               sse_algorithm = "AES256"
                           }
                       }
                   }

                   versioning {
                       enabled    = true
                       mfa_delete = false
                   }
               }

           Plan: 1 to import, 0 to add, 1 to change, 0 to destroy.

           ─────────────────────────────────────────────────────────────────────────────

           Saved the plan to: plan

           To perform exactly these actions, run the following command to apply:
               terraform apply "plan"
```

### Importing without Configuration

If you want to import a resource without specifying the configuration you can use the instance method `import` on the class of the resource you want to import. This method takes the scope as the first argument, the construct name within the CDKTF program as the second, resource ID as the third argument and the provider as an optional fourth. The provider is only required if you have multiple providers of the same type in your configuration.

```typescript

> S3Bucket.import(this, "bucket", bucketId);

Reviewer: this method has been renamed, hasn't it?

```

> When running plan / apply you will get the information that one resource is going to be imported. Once you ran plan / apply, CDKTF CLI will return the CDKTF construct. You can use this in your CDKTF application after the the apply and therefore the import is done and replace the `S3Bucket.import` call.

Reviewer: This needs some updating since `apply` will fail if the config is not yet present. I suspect the convert output has to be put into the config first (with an importFrom as specified above) and then an apply is needed.

                              import { Construct } from "constructs";
                              /*
                               * Provider bindings are generated by running `cdktf get`.
                               * See https://cdk.tf/provider-generation for more details.
                               */
                              import { S3Bucket } from "./.gen/providers/aws/s3-bucket";
                              class MyConvertedCode extends Construct {
                                constructor(scope: Construct, name: string) {
                                  super(scope, name);
                                  new S3Bucket(this, "bucket", {
                                    bucket: "best-bucket-in-the-world",
                                    bucketPrefix: [null],
                                    forceDestroy: [null],
                                    objectLockEnabled: false,
                                    tags: {
                                      foo: "bar",
                                    },
                                    tagsAll: {
                                      foo: "bar",
                                    },

>                                   });

Reviewer: This could be a follow-up issue as it would require changes to convert, but adding this code to the code base is not really enough – it would still require an `.importFrom()` call on the bucket to be able to apply after generating.

If you need to ensure a condition is met either before or after a resource was created you can specify [conditions](/terraform/language/expressions/custom-conditions#preconditions-and-postconditions).
To add one configure the `lifecycle` key on your resource with an object containing a `precondition` and / or a `postcondition`. These keys take a list of conditions with a `condition` key containing a Terraform Expression to be evaluated and an `errorMessage` key containing a string to be displayed if the condition is not met.

## Importing Resources

If you have existing resources that you want to manage with CDKTF, you can import them into your CDKTF application. The best way to do this is using the [`import` block feature](/terraform/language/import) of Terraform >= 1.5. You can do this in CDKTF either with a specified configuration or without.

### How To Import

> To import a resource, first instantiate an instance of the resource type you wish to import– in our case we'll be using a S3Bucket. No configuration is expilictly needed. You then call the `importFrom` method on the resource object. This method takes the ID of the resource to be imported as the first argument and the provider as an optional second. The provider is only required if you have multiple providers of the same type in your configuration.

Reviewer 2: Good reminder to have spell check when I'm writing docs 😅. Included 👍

```typescript
new S3Bucket(this, "bucket", {}).importFrom(bucketId);
```

> When running plan / apply you will get the information that your resource is going to be imported. Once you have ran apply, you can remove the `importFrom` call and the resource will become managed by CDKTF.

Reviewer: ```suggestion
When running plan / apply you will get the information that your resource is going to be imported. Once you have ran apply, you can remove the `importFrom` call and the resource will become managed by CDKTF.

Please note that Terraform is going to update existing fields on the imported resource to match your configuration as it puts it under management. In our case we did not define any specific properties on the `S3Bucket` which causes Terraform e.g. to remove the tags currently defined on the resource (as can be seen on the plan below). If you want to keep existing settings, you can run a plan first, add everything that Terraform would change to your resource config, and only then apply the changes.
```

Reviewer 2: included 👍

                     - "foo" = "bar"
                   } -> (known after apply)

                   grant {
                       id          = "XXXXXXXXXXXXX"
                       permissions = [
                           "FULL_CONTROL",
                       ]
                       type        = "CanonicalUser"
                   }

                   server_side_encryption_configuration {
                       rule {
                           bucket_key_enabled = true

                           apply_server_side_encryption_by_default {
                               sse_algorithm = "AES256"
                           }
                       }
                   }

                   versioning {
                       enabled    = true
                       mfa_delete = false
                   }
               }

           Plan: 1 to import, 0 to add, 1 to change, 0 to destroy.

           ─────────────────────────────────────────────────────────────────────────────

           Saved the plan to: plan

           To perform exactly these actions, run the following command to apply:
               terraform apply "plan"
```

### Generate Configuration For Import

> If you want to specify the configuration of your imported resource you can use the static method `generateConfigForImport` on the class of the resource you want to import. This method takes the scope as the first argument, the resource ID of the resource to import to (as will be given in the generated config returned), the resource ID of the resource to be imported, and the provider as an optional fourth. The provider is only required if you have multiple providers of the same type in your configuration.

Reviewer: ```suggestion
If you don't want to specify the configuration of your imported resource yourself you can use the static method `generateConfigForImport` on the class of the resource you want to import. This method takes the scope as the first argument, the construct id of the resource to import to (as will be given in the generated config returned), the resource id of the resource to be imported, and the provider as an optional fourth. The provider is only required if you have multiple providers of the same type in your configuration.
```

Reviewer 2: included 👍

```typescript
S3Bucket.generateConfigForImport(this, "bucket", bucketId);
```

> When running `cdktf plan <stack-name>` you will get the information that resource(s) is/are going to be imported, followed by the configuration of your import(s) in CDKTF specific terms.

Reviewer: I found the original sentence to be a bit too abstract, dunno if this is better though:
```suggestion
When running `cdktf plan <stack-name>` Terraform will generate code for the resource you are importing and CDKTF will convert it to the language you are using.
```

                               */
                              import { S3Bucket } from "./.gen/providers/aws/s3-bucket";
                              class MyConvertedCode extends Construct {
                                constructor(scope: Construct, name: string) {
                                  super(scope, name);
                                  new S3Bucket(this, "bucket", {
                                    bucket: "best-bucket-in-the-world",
                                    bucketPrefix: [null],
                                    forceDestroy: [null],
                                    objectLockEnabled: false,
                                    tags: {
                                      foo: "bar",
                                    },
                                    tagsAll: {
                                      foo: "bar",
                                    },
                                  });
                                }
                              }

                              Please review the code and make any necessary changes before adding it to your codebase.

>                               Make sure to only copy the code within the construct's constructor.

Reviewer: We need to update this to match what's printed now 🙈

```

> Though at this point, your resource(s) has/have not been imported. To import, first add the new generated configuration to your project, then remove the initial call of `generateConfigForImport`. Finally, add the call `importFrom` as it is done in the section "How To Import" above. On apply, your resource(s) will be imported, then becoming managed by CDKTF.

Reviewer: We could reference the `How to import` section above as it has the details to follow from here on 🤔

Reviewer 2: Shortened what I had a bit and referenced the `How To Import` section more strongly 👍
